In [1]:
!pip install -q nltk
import nltk
nltk.download("book")
nltk.download("punkt")
from nltk.tokenize import word_tokenize
from nltk.tokenize import sent_tokenize

[nltk_data] Downloading collection 'book'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/abc.zip.
[nltk_data]    | Downloading package brown to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/brown.zip.
[nltk_data]    | Downloading package chat80 to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/chat80.zip.
[nltk_data]    | Downloading package cmudict to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/cmudict.zip.
[nltk_data]    | Downloading package conll2000 to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/conll2000.zip.
[nltk_data]    | Downloading package conll2002 to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/conll2002.zip.
[nltk_data]    | Downloading package dependency_treebank to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping corpora/dependency_treebank.zip.
[nltk_data]    | Downloading package genesis to /root/nltk_data...
[nltk_data]    

#Part 1

##Part A
###Tokenize all sentences (use sent_tokenize), and extract all tokens from each sentence (use word_tokenize). Also lowercase all tokens. Do not use other processing steps for now. Keep the tokens in each sentence in a list (you should have a list of lists of tokens). Display the number of sentences you found.

In [2]:
with open("/content/alice.txt", 'r') as f:
  text = f.read()  # Read the full text and store as a string

(1) Concatenate all tokens from the list of lists at (A) in a single list of tokens. Display the number of tokens.

In [3]:
sentences = nltk.sent_tokenize(text) # tokenizes all sentences in text
print(f"Number of sentences found: {len(sentences)}") # prints number of sentences
tokens_sentence = [nltk.word_tokenize(sentence.lower()) for sentence in sentences] # tokenzies the words for each sentence and lowercase the tokens
all_tokens = sum(tokens_sentence, []) # concatenates lists of tokens from tokens_sentence into one list
print(f"Number of tokens found: {len(all_tokens)}") # prints number of tokens
print(all_tokens[:20])

Number of sentences found: 1616
Number of tokens found: 34399
['alice', "'s", 'adventures', 'in', 'wonderland', 'alice', "'s", 'adventures', 'in', 'wonderland', 'lewis', 'carroll', 'the', 'millennium', 'fulcrum', 'edition', '3.0', 'chapter', 'i', 'down']


####(2) Write a function that stores the unique tokens and their frequency in a dictionary (write your own function).
####The function needs to return a dictionary with all tokens as keys and frequencies as values. Display the number of unique tokens.


In [4]:

def token_dictionary(tokens):
  freq_dict = {} # dictionary that stores frequency of each token
  for token in tokens:
    if token in freq_dict:
      freq_dict[token] += 1 # if token in dictionary increment count by 1
    else:
      freq_dict[token] = 1 # if token not in dictionary add it
  return freq_dict

In [5]:
token_freq_dict = token_dictionary(all_tokens) # stores dictionary of tokens
print(f"Number of unique tokens: {len(token_freq_dict)}")

Number of unique tokens: 2650


####(3) Write a function that sorts a dictionary by values from largest to lowest.
####In this case, the values are frequencies (see example at the end of notebook_intro_nlp.ipynb in class repository).
####Sort the dictionary at A.2. and display the 10 most and the 10 least frequent terms and their frequencies.


In [6]:
def sort_by_frequency(freq_dict):
    sorted_freq = sorted(freq_dict.items(), key=lambda item: item[1], reverse=True) # sorts the dictionary by values in descending order
    return sorted_freq

In [8]:
sorted_token_freq = sort_by_frequency(token_freq_dict) #  sorts the token frequency dictionary

print("10 most frequent terms:")
for token, freq in sorted_token_freq[:10]: # starts listing the top 10 frequency tokens
    print(f"{token}: {freq}")

print("\n10 least frequent terms:")
for token, freq in sorted_token_freq[-10:]: # starts from the end of the sorted dictionary and displays the least 10 frequency tokens
    print(f"{token}: {freq}")

10 most frequent terms:
,: 2418
the: 1639
`: 1109
': 1102
.: 970
and: 866
to: 725
a: 631
it: 591
she: 553

10 least frequent terms:
riper: 1
years: 1
loving: 1
childhood: 1
gather: 1
sorrows: 1
joys: 1
remembering: 1
child-life: 1
happy: 1


####(4) Remove all punctuation marks from the dictionary at (3).
####Display the number of unique tokens, the 10 most and 10 least frequent terms and their frequencies.

In [9]:
import re

def remove_punctuation(tokens): # function to remove punctuation from tokens
    return [re.sub(r'[^\w\s]', '', token) for token in tokens if re.sub(r'[^\w\s]', '', token)]

In [10]:
all_tokens = remove_punctuation(all_tokens)
token_freq_dict = token_dictionary(all_tokens) # stores dictionary of tokens without punctuation

print(f"Number of unique tokens (without punctuation marks): {len(token_freq_dict)}")

sorted_token_freq = sort_by_frequency(token_freq_dict) # sorts the token frequency dictionary

print("10 most frequent terms (without punctuation marks):")
for token, freq in sorted_token_freq[:10]: # starts listing the top 10 frequency tokens
    print(f"{token}: {freq}")

print("\n10 least frequent terms (without punctuation marks):")
for token, freq in sorted_token_freq[-10:]: # starts from the end of the sorted dictionary and displays the least 10 frequency tokens
    print(f"{token}: {freq}")

Number of unique tokens (without punctuation marks): 2615
10 most frequent terms (without punctuation marks):
the: 1639
and: 866
to: 725
a: 631
it: 591
she: 553
i: 525
of: 511
said: 462
you: 406

10 least frequent terms (without punctuation marks):
riper: 1
years: 1
loving: 1
childhood: 1
gather: 1
sorrows: 1
joys: 1
remembering: 1
childlife: 1
happy: 1


####(5) Remove all stop words from the dictionary at (4).
####Display the number of unique tokens, the 10 most and 10 least frequent terms and their frequencies.

In [11]:
from nltk.corpus import stopwords
def remove_stopwords(tokens): # function that removes stopwords
    stop_words = set(stopwords.words('english')) # sets all words that are considered stop words in strings
    return [token for token in tokens if token not in stop_words]

In [12]:
all_tokens = remove_stopwords(all_tokens)
token_freq_dict = token_dictionary(all_tokens) # stores dictionary of tokens without punctuation

print(f"Number of unique tokens (without stopwords): {len(token_freq_dict)}")

sorted_token_freq = sort_by_frequency(token_freq_dict) # sorts the token frequency dictionary

print("10 most frequent terms (without stopwords):")
for token, freq in sorted_token_freq[:10]: # starts listing the top 10 frequency tokens
    print(f"{token}: {freq}")

print("\n10 least frequent terms (without stopwords):")
for token, freq in sorted_token_freq[-10:]: # starts from the end of the sorted dictionary and displays the least 10 frequency tokens
    print(f"{token}: {freq}")

Number of unique tokens (without stopwords): 2483
10 most frequent terms (without stopwords):
said: 462
alice: 399
nt: 208
little: 128
one: 104
would: 96
know: 88
could: 86
like: 85
went: 83

10 least frequent terms (without stopwords):
riper: 1
years: 1
loving: 1
childhood: 1
gather: 1
sorrows: 1
joys: 1
remembering: 1
childlife: 1
happy: 1


####(6) Stem all remaining words.
####Use the PorterStemmer and rebuild the dictionary so that identical stemmed words appear only once and their frequencies are summed.
####Display the number of unique tokens, the 10 most and 10 least frequent terms and their frequencies.

In [13]:
from nltk.stem import PorterStemmer
stemmer = PorterStemmer() # initalizes the Porter Stemmer
def stem_tokens(tokens): # function to stem tokens using Porter Stemmer
    return [stemmer.stem(token) for token in tokens]

In [14]:
all_tokens = stem_tokens(all_tokens)
token_freq_dict = token_dictionary(all_tokens) # stores dictionary of tokens without punctuation

print(f"Number of unique tokens (without stopwords): {len(token_freq_dict)}")

sorted_token_freq = sort_by_frequency(token_freq_dict) # sorts the token frequency dictionary

print("10 most frequent terms (without stopwords):")
for token, freq in sorted_token_freq[:10]: # starts listing the top 10 frequency tokens
    print(f"{token}: {freq}")

print("\n10 least frequent terms (without stopwords):")
for token, freq in sorted_token_freq[-10:]: # starts from the end of the sorted dictionary and displays the least 10 frequency tokens
    print(f"{token}: {freq}")

Number of unique tokens (without stopwords): 1899
10 most frequent terms (without stopwords):
said: 462
alic: 399
nt: 208
littl: 128
one: 105
look: 104
like: 97
would: 96
know: 92
could: 86

10 least frequent terms (without stopwords):
farmyard: 1
cattl: 1
lastli: 1
aftertim: 1
riper: 1
childhood: 1
gather: 1
joy: 1
childlif: 1
happi: 1


##Part B
###Use the list of sentence tokens you obtained at 1.A


### (1) Extract the POS (Parts Of Speech) of all tokens in each sentence (use pos_tag from nltk library).You will obtain a list of tuples for each sentence. All sentence lists will be inside a list. A tuple will contain the token and its POS. Display the tuples in a few sentences in the middle of the book.

In [15]:
from nltk import pos_tag

def extract_pos(tokens_sentence): # function to extract POS of all tokens in each sentence
  pos_sentences = [pos_tag(tokens) for tokens in tokens_sentence] # gets the POS of each sentence
  return pos_sentences

In [16]:
pos_sentences = extract_pos(tokens_sentence) # extracts POS of all tokens for each sentence
middle_index = len(pos_sentences) // 2 # finds the middle index of the text
print("\nPOS for sentences in the middle of the book: \n")

for sentence_pos in pos_sentences [middle_index:middle_index + 5]: # displays the 5 sentences in the middle of the book and their POS
  print(sentence_pos)
  print("\n")


POS for sentences in the middle of the book: 

[('`', '``'), ('i', 'NN'), ('dare', 'NN'), ('say', 'VBP'), ('you', 'PRP'), ('never', 'RB'), ('even', 'RB'), ('spoke', 'VBD'), ('to', 'TO'), ('time', 'NN'), ('!', '.'), ("'", "''")]


[('`', '``'), ('perhaps', 'RB'), ('not', 'RB'), (',', ','), ("'", "''"), ('alice', 'RB'), ('cautiously', 'RB'), ('replied', 'VBN'), (':', ':'), ('`', '``'), ('but', 'CC'), ('i', 'NN'), ('know', 'VBP'), ('i', 'RB'), ('have', 'VBP'), ('to', 'TO'), ('beat', 'VB'), ('time', 'NN'), ('when', 'WRB'), ('i', 'NN'), ('learn', 'VBP'), ('music', 'NN'), ('.', '.'), ("'", "''")]


[('`', '``'), ('ah', 'NN'), ('!', '.')]


[('that', 'DT'), ('accounts', 'VBZ'), ('for', 'IN'), ('it', 'PRP'), (',', ','), ("'", "''"), ('said', 'VBD'), ('the', 'DT'), ('hatter', 'NN'), ('.', '.')]


[('`', '``'), ('he', 'PRP'), ('wo', 'MD'), ("n't", 'RB'), ('stand', 'VB'), ('beating', 'NN'), ('.', '.')]




### (2) Lemmatize the nouns, verbs, adjectives, and adverbs in each sentence using nltk (see notebook_nlp_intro.ipynb): noun tokens have POS that starts with N, verb tokens, have POS that starts with V, adverb tokens have POS that starts with R, and adjective tokens have POS that starts with J. You will have to change the original POS into a letter 'n' for nouns, 'v' for verbs, 'a' for adjectives, and 'r' for adverbs. Do not remove any other tokens for now. Display the tuples in a few sentences in the middle of the book.

In [17]:
from nltk.stem import WordNetLemmatizer
from nltk.corpus import wordnet

lemmatizer = WordNetLemmatizer() # initializes the WordNet Lemmatizer

def get_wordnet(tag): # function that maps nltk POS tags to wordnet POS tags
  if tag.startswith('N'):
    return wordnet.NOUN
  if tag.startswith('V'):
    return wordnet.VERB
  if tag.startswith('J'):
    return wordnet.ADJ
  if tag.startswith('R'):
    return wordnet.ADV
  else:
    return None

In [18]:
def lemmatize_sentence(pos_sentences):
  lemmatized_sentence = []
  for word, pos in pos_sentences:
    wordnet_pos = get_wordnet(pos) # gets the wordnet POS
    if wordnet_pos: # if the word POS is a noun, verb, adjective or adverb
      lemmatized_word = lemmatizer.lemmatize(word, wordnet_pos)
    else:
      lemmatized_word = word # keeps the word as is if not Noun, Verb, Adjective, or Adverb
    lemmatized_sentence.append((lemmatized_word, pos)) # keeps the original POS
  return lemmatized_sentence

In [19]:
lemmatized_sentences = [lemmatize_sentence(sentence) for sentence in pos_sentences] # lemmatizes each sentence

middle_index = len(lemmatized_sentences) // 2 # finds the middle index of the text
print("\n Lemmatized sentences in the middle of the book: \n")

for sentence in lemmatized_sentences [middle_index:middle_index + 5]: # displays the 5 sentences in the middle of the book and their POS
  print(sentence)
  print("\n")


 Lemmatized sentences in the middle of the book: 

[('`', '``'), ('i', 'NN'), ('dare', 'NN'), ('say', 'VBP'), ('you', 'PRP'), ('never', 'RB'), ('even', 'RB'), ('speak', 'VBD'), ('to', 'TO'), ('time', 'NN'), ('!', '.'), ("'", "''")]


[('`', '``'), ('perhaps', 'RB'), ('not', 'RB'), (',', ','), ("'", "''"), ('alice', 'RB'), ('cautiously', 'RB'), ('reply', 'VBN'), (':', ':'), ('`', '``'), ('but', 'CC'), ('i', 'NN'), ('know', 'VBP'), ('i', 'RB'), ('have', 'VBP'), ('to', 'TO'), ('beat', 'VB'), ('time', 'NN'), ('when', 'WRB'), ('i', 'NN'), ('learn', 'VBP'), ('music', 'NN'), ('.', '.'), ("'", "''")]


[('`', '``'), ('ah', 'NN'), ('!', '.')]


[('that', 'DT'), ('account', 'VBZ'), ('for', 'IN'), ('it', 'PRP'), (',', ','), ("'", "''"), ('say', 'VBD'), ('the', 'DT'), ('hatter', 'NN'), ('.', '.')]


[('`', '``'), ('he', 'PRP'), ('wo', 'MD'), ("n't", 'RB'), ('stand', 'VB'), ('beating', 'NN'), ('.', '.')]




### (3) Write a function that stores the unique tuples (token,POS) and their frequency in a dictionary (write your own function). The function needs to return a dictionary with all unique tuples as keys and frequencies as values. Display the number of unique tuples.

In [20]:
def tuple_frequency_dict(pos_sentences):
  freq_dict = {} # stores tuple frequency
  for sentence in pos_sentences: # iterates over POS sentences
    for token_pos_tuple in sentence: # iterates over each tuple in the sentence
      if token_pos_tuple in freq_dict:
        freq_dict[token_pos_tuple] +=1 # increments the count if the tuple is already in the dictionary
      else:
        freq_dict[token_pos_tuple] = 1 # adds the tuple to the dictionary with a count of 1
  return freq_dict

In [21]:
tuple_freq_dict = tuple_frequency_dict(lemmatized_sentences)
print(f"Number of unique tuples: {len(tuple_freq_dict)}")

Number of unique tuples: 3244


### (4) Sort the dictionary by frequencies from largest to smallest tuple (see example at the end of notebook_intro_nlp.ipynb in class repository). Display the 10 most and the 10 least frequent tuples and their frequencies.

In [22]:
sorted_tuple_freq = sort_by_frequency(tuple_freq_dict) #  sorts the tuple frequency dictionary

print("10 most frequent tuples:")
for token_pos_tuple, freq in sorted_tuple_freq[:10]: # starts listing the top 10 tuple frequency
    print(f"{token_pos_tuple}: {freq}")

print("\n10 least frequent tuples:")
for token_pos_tuple, freq in sorted_tuple_freq[-10:]: # starts from the end of the sorted dictionary and displays the least 10 tuple frequency tokens
    print(f"{token_pos_tuple}: {freq}")

10 most frequent tuples:
(',', ','): 2418
('the', 'DT'): 1639
('`', '``'): 1109
("'", "''"): 1061
('.', '.'): 970
('and', 'CC'): 866
('to', 'TO'): 725
('a', 'DT'): 631
('it', 'PRP'): 591
('she', 'PRP'): 553

10 least frequent tuples:
('love', 'VBG'): 1
('childhood', 'NN'): 1
('gather', 'VB'): 1
('bright', 'VBN'): 1
('ago', 'NN'): 1
('sorrow', 'NNS'): 1
('joy', 'NNS'): 1
('remember', 'VBG'): 1
('child-life', 'NN'): 1
('happy', 'JJ'): 1


### (5) Remove all tuples with empty POS (it should remove all punctuation marks). Display the number of unique tuples and the 10 most and 10 least frequent tuples and their frequencies.

In [23]:
def remove_punct_tuples(freq_dict):
  return {key: value for key, value in freq_dict.items() if key[1].isalpha()} # only keeps tuples that are alphabetical, so punctuation marks are removed

tuple_freq_dict_npunct = remove_punct_tuples(tuple_freq_dict) # removes punctuation marks from tuples
print(f"Number of unique tuples with no punctuation marks: {len(tuple_freq_dict_npunct)}")

sorted_tuple_freq_npunct = sort_by_frequency(tuple_freq_dict_npunct) # sorts the updated dictionary

print("10 most frequent tuples with no punctuation marks:")
for token_pos_tuple, freq in sorted_tuple_freq_npunct[:10]: # starts listing the top 10 tuple frequency with no punctuation
    print(f"{token_pos_tuple}: {freq}")

print("\n10 least frequent tuples with no punctuation marks:")
for token_pos_tuple, freq in sorted_tuple_freq_npunct[-10:]: # starts from the end of the sorted dictionary and displays the least 10 tuple frequency tokens with no punctuation
    print(f"{token_pos_tuple}: {freq}")

Number of unique tuples with no punctuation marks: 3221
10 most frequent tuples with no punctuation marks:
('the', 'DT'): 1639
('and', 'CC'): 866
('to', 'TO'): 725
('a', 'DT'): 631
('it', 'PRP'): 591
('she', 'PRP'): 553
('of', 'IN'): 511
('say', 'VBD'): 462
('be', 'VBD'): 454
('you', 'PRP'): 406

10 least frequent tuples with no punctuation marks:
('love', 'VBG'): 1
('childhood', 'NN'): 1
('gather', 'VB'): 1
('bright', 'VBN'): 1
('ago', 'NN'): 1
('sorrow', 'NNS'): 1
('joy', 'NNS'): 1
('remember', 'VBG'): 1
('child-life', 'NN'): 1
('happy', 'JJ'): 1


### (6) Remove all tuples that are NOT nouns, adjectives, adverbs, or verbs. Display the number of unique tuples and the 10 most and 10 least frequent tuples and their frequencies.

In [24]:
def remove_naav_tuples(freq_dict): # function that remove tuples that are not Nouns, Adjectives, Adverbs, Verbs
  return {key: value for key, value in freq_dict.items() if key[1] == "NN" or key[1] == "JJ" or key[1] == "RB" or key[1] == "VB"} # only keeps Nouns, Adjectives, Adverbs, Verbs

tuple_freq_dict_naav = remove_naav_tuples(tuple_freq_dict_npunct) # removes tuples that are not Nouns, Adjectives, Adverbs, or Verbs
print(f"Number of unique tuples with only Nouns, Adj, Adverbs, and Verbs: {len(tuple_freq_dict_naav)}")

sorted_tuple_freq_naav = sort_by_frequency(tuple_freq_dict_naav) # sorts the updated dictionary

print("10 most frequent tuples with only Nouns, Adj, Adverbs, and Verbs:")
for token_pos_tuple, freq in sorted_tuple_freq_naav[:10]: # starts listing the top 10 tuple frequency with only Nouns, Adjectives, Adverbs, Verbs
    print(f"{token_pos_tuple}: {freq}")

print("\n10 least frequent tuples with only Nouns, Adj, Adverbs, and Verbs:")
for token_pos_tuple, freq in sorted_tuple_freq_naav[-10:]: # starts from the end of the sorted dictionary and displays the least 10 tuple frequency tokens with only Nouns, Adjectives, Adverbs, Verbs
    print(f"{token_pos_tuple}: {freq}")

Number of unique tuples with only Nouns, Adj, Adverbs, and Verbs: 1888
10 most frequent tuples with only Nouns, Adj, Adverbs, and Verbs:
('alice', 'NN'): 311
('i', 'NN'): 276
("n't", 'RB'): 208
('be', 'VB'): 148
('not', 'RB'): 146
('very', 'RB'): 144
('little', 'JJ'): 127
('so', 'RB'): 118
('i', 'JJ'): 98
('then', 'RB'): 94

10 least frequent tuples with only Nouns, Adj, Adverbs, and Verbs:
("turtle's", 'NN'): 1
('lastly', 'RB'): 1
('after-time', 'JJ'): 1
('grown', 'JJ'): 1
('riper', 'NN'): 1
('childhood', 'NN'): 1
('gather', 'VB'): 1
('ago', 'NN'): 1
('child-life', 'NN'): 1
('happy', 'JJ'): 1


### 1(C) - As I added more processing steps to the dictionary of words from the Alice in Wonderland text, the number of unique tokens goes down for when I went from A.1 to A.6 and B.1 to B.6. Immediately, the steps through 1B end up giving a more informative dictionary, since it gives the part of speech of the word on top of the word frequency in the text. In 1A.2, I had 2650 unique tokens and in 1B.3 I had 3244 unique tokens, and although the number of tokens does decrease by a great amount, the steps through 1.B help lead to a better list of words that is shorter and much more informative than doing the steps through 1.A. This is due to the fact that the stop words removal with stemming  ends up making some words hard for identifying what part of speech they are, thus making it harder to  identify the word in general, whereas the POS removal with lemmatization is much more efficient since the final count of unique tokens is 1888 for 1.B which is less than the count in 1.A and it also contains more important terms since its only the nouns, adjectives, adverbs, and verbs from the text.

### (D) Propose at least one additional processing step for each (A) and (B) that you think will improve the resulting list of tokens - shorter and more informative of the subject matter. Implement it, show its effect and comment on the results you obtained.

In [25]:
def remove_short_tokens(freq_dict, min_length=3):
    return {token: freq for token, freq in freq_dict.items() if len(token) >= min_length} # removes words that are shorter than 2 characters

token_freq_dict_st = remove_short_tokens(token_freq_dict, min_length=3) # removes short tokens

sorted_token_freq_st = sort_by_frequency(token_freq_dict_st) # sorts and displays the updated token frequency dictionary

print(f"Number of unique tokens after filtering short tokens: {len(token_freq_dict_st)}")

print("10 most frequent tokens (filtered by length):") # displays the 10 most frequent tokens after removing short tokens
for token, freq in sorted_token_freq_st[:10]:
    print(f"{token}: {freq}")

print("\n10 least frequent tokens (filtered by length):") # displays 10 least frequent tokens after removing short tokens
for token, freq in sorted_token_freq_st[-10:]:
    print(f"{token}: {freq}")

Number of unique tokens after filtering short tokens: 1870
10 most frequent tokens (filtered by length):
said: 462
alic: 399
littl: 128
one: 105
look: 104
like: 97
would: 96
know: 92
could: 86
went: 83

10 least frequent tokens (filtered by length):
farmyard: 1
cattl: 1
lastli: 1
aftertim: 1
riper: 1
childhood: 1
gather: 1
joy: 1
childlif: 1
happi: 1


In [26]:
def remove_low_frequency_tuples(freq_dict, min_freq=2): # function removes tuples with low frequencies like 2 or less
    return {key: value for key, value in freq_dict.items() if value > min_freq}

tuple_freq_dict_naav = remove_low_frequency_tuples(tuple_freq_dict_naav, min_freq=2) # removes low-frequency tuples

sorted_tuple_freq_naav = sort_by_frequency(tuple_freq_dict_naav) # sorts and displays the updated tuple frequency dictionary

print(f"Number of unique tuples after low-frequency tuple removal: {len(sorted_tuple_freq_naav)}")

print("10 most frequent tuples after low-frequency tuple removal:") # displays the 10 most frequent tuples after low-frequency tuple removal
for token_pos_tuple, freq in sorted_tuple_freq_naav[:10]:
    print(f"{token_pos_tuple}: {freq}")

print("\n10 least frequent tuples after low-frequency tuple removal:") # displays the 10 least frequent tuples after low-frequency tuple removal
for token_pos_tuple, freq in sorted_tuple_freq_naav[-10:]:
    print(f"{token_pos_tuple}: {freq}")

Number of unique tuples after low-frequency tuple removal: 645
10 most frequent tuples after low-frequency tuple removal:
('alice', 'NN'): 311
('i', 'NN'): 276
("n't", 'RB'): 208
('be', 'VB'): 148
('not', 'RB'): 146
('very', 'RB'): 144
('little', 'JJ'): 127
('so', 'RB'): 118
('i', 'JJ'): 98
('then', 'RB'): 94

10 least frequent tuples after low-frequency tuple removal:
('quadrille', 'NN'): 3
('figure', 'NN'): 3
('snail', 'NN'): 3
('owl', 'NN'): 3
('panther', 'NN'): 3
('trumpet', 'NN'): 3
('consider', 'VB'): 3
('teacup', 'NN'): 3
('list', 'NN'): 3
('swim', 'VB'): 3


### 1(D) - For steps A, I implemented an additional processing step that removes short tokens from the list of tokens, in which if the token only has a length of 2 characters then it will be removed. I did this since tokens that only have 1-2 characters are terms such as "a" or "is" and ultimately they are not that informative on the overall subject matter of the text "Alice in Wonderland". Once I implemented this, even though the count of unique tokens did not go down by that much, it still resulted in a more informative list of terms. For steps B, I implemented an additional processing step that removes tuples that have a very small frequency, in which it is a term that only shows up 1-2 times in the text. I did this since a term that only shows up 1-2 times in the text is most definitely not necessary to understand the overall subject matter. Once I implemented this, I got a very concise list of tuples where there were only 645 tuples, all with their respective parts of speech, so its a very short but informative list on the subject matter.

# Part 2

### (A) Use spacy to process the file and to apply the processing steps (lowercase, remove stopwords and punctuation marks, lemmatize words). Display the number of unique tokens, the 10 most and 10 least frequent terms, and their frequencies (use the Counter object in Python)

In [27]:
!pip install -q spacy
!python -m spacy download en_core_web_sm

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.8/12.8 MB 84.4 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


In [36]:
import spacy
from collections import Counter

nlp = spacy.load("en_core_web_sm")

doc_processed = nlp(text.lower()) # processes the text and lowercases all characters

tokens_spac = [token.lemma_ for token in doc_processed if not token.is_stop and not token.is_punct and token.lemma_.isalpha()] # creates a list to store the lemmatized tokens that also removes stopwords and punctuation marks

token_frequency = Counter(tokens_spac) # token frequency is calculated

print(f"Number of unique tokens: {len(token_frequency)}") # displays number of unique tokens

sorted_token_freq = token_frequency.most_common() # sorts tokens by frequency

print("10 most frequent terms:") # displays the 10 most frequent terms
for token, freq in sorted_token_freq[:10]:
    print(f"{token}: {freq}")

print("\n10 least frequent terms:") # displays the 10 least frequent terms
for token, freq in sorted_token_freq[-10:]:
    print(f"{token}: {freq}")

Number of unique tokens: 1765
10 most frequent terms:
say: 475
alice: 398
think: 129
go: 129
little: 127
look: 105
know: 103
like: 92
begin: 91
come: 90

10 least frequent terms:
lowing: 1
cattle: 1
lastly: 1
grown: 1
woman: 1
riper: 1
childhood: 1
gather: 1
joy: 1
happy: 1


### 2(B) - Each processing step goes much faster and much more efficiently through using spacy, in which the code that is required to get the processing steps done is very concise sue to the fact that spacy has built in functions such as is_stop and is_punct to check if a word is a stopword or a punctuation mark. Compared to the results that I got from part 1 with nltk processing, spacy proved that it ends up making a shorter and informative list since it almost has 100 less tokens than both parts A and B with a count of 1750 unique tokens through spacy, and it definitely could be more informative if it was given more processing steps.

### (C) Read more about spacy and based on its features propose an additional processing step that will improve the final list of tokens. Implement it, and comment on the results you obtained.

In [33]:
tokens_spac2 = [] # creates a list that will sotre the lemmatized tokens that will only contain named entities and nouns, verbs, adjectives, adverbs

for token in doc_processed:
  if token.ent_type_ or token.pos_ in ["NOUN", "VERB", "ADJ", "ADV"]: # checks if token is named entitiy or noun, verb, adj, adverb
    if not token.is_stop and not token.is_punct and token.lemma_.isalpha(): # checks if token is stopword, punctuation mark, or non-alphabetical
      tokens_spac2.append(token.lemma_) # stores lemmatized token in list

token_freq = Counter(tokens_spac2) # token frequency is calculated

print(f"Number of unique tokens with only named entities, nouns, adj, verbs, adverbs: {len(token_freq)}") # displays number of unique tokens with only named entities and nouns, adj, verbs, adverbs

sorted_token_freq2 = token_freq.most_common() # sorts tokens by frequency

print("10 most frequent terms with only named entities, nouns, adj, verbs, adverbs:") # displays the 10 most frequent terms
for token, freq in sorted_token_freq2[:10]:
    print(f"{token}: {freq}")

print("\n10 least frequent terms with only named entities, nouns, adj, verbs, adverbs:") # displays the 10 least frequent terms
for token, freq in sorted_token_freq2[-10:]:
    print(f"{token}: {freq}")

Number of unique tokens with only named entities, nouns, adj, verbs, adverbs: 1708
10 most frequent terms with only named entities, nouns, adj, verbs, adverbs:
say: 475
think: 129
go: 129
little: 127
look: 105
know: 103
begin: 91
come: 89
thing: 79
alice: 78

10 least frequent terms with only named entities, nouns, adj, verbs, adverbs:
lowing: 1
cattle: 1
lastly: 1
grown: 1
woman: 1
riper: 1
childhood: 1
gather: 1
joy: 1
happy: 1


### 2(C) - After reading more about spacy and the features that it has I decided to implement NER as the additional processing step since by using NER, it can extract important entities such as important names of people and places in the text "Alice in Wonderland" and these entities will end up making the list much more informative. Thus, I decided to keep only named entities and terms that are nouns, adjectives, verbs, and adverbs, since I believe will make a list that contains most words that are important to the subject matter of "Alice in Wonderland". After implementing this, I saw that the results were a shorter and much more informative list that can help identify what the text is much faster than the other lists through part 1.